In [217]:
import pandas as pd
import collections
from utils import *

In [218]:
OUTDIR = 'output'
INDIR_BILL = './data/bills' 
INDIR_SUMMARY = './data/summary' 

In [ ]:
# version_def = pd.read_csv('version_definition.csv')
# version_def['Version Abbreviation'] = version_def['Version Abbreviation'].str.upper()

In [199]:
# bills = []

# for root, dirs, filenames in os.walk(INDIR_BILL):
#     for filename in filenames:
#         b_type, b_subtype, b_number, b_version, b_filename = extract_nameinfo(filename)
        
#         with open(os.path.join(INDIR_BILL, filename), 'rb') as f:
#             doc = xmltodict.parse(f.read())
            
#             b_tags = [key for key, value in recursive_items(doc)]
#             b_tags = list(set(b_tags))
            
#             b_dict = {'Type': b_type, 
#                       'Subtype': b_subtype, 
#                       'Number': b_number,
#                       'Version Abbreviation': b_version,
#                       'Filename': b_filename,
#                       'Tags': b_tags
#                      }
            
#             bills.append(b_dict)
            
# df_bills = pd.DataFrame(bills)
# df_bills = pd.merge(df_bills, version_def, how='left', on='Version Abbreviation', suffixes=('', '_'))
# df_bills.to_csv('tags.csv', index=False)

# Bills

In [222]:
df_bills = pd.read_csv('tags.csv')

In [304]:
len(df_bills)

8039

In [223]:
df_bills.groupby(['Type']).count()[['Filename']].rename(columns={'Filename': 'Count'})

Count
Type                         
Bills                    6726
Concurrent Resolutions    147
Joint Resolutions         241
Simple Resolutions        924

In [224]:
df_bills.groupby(['Type', 'Subtype']).count()[['Filename']].rename(columns={'Filename': 'Count'})

Count
Type                   Subtype       
Bills                  HR        4639
                       S         2087
Concurrent Resolutions HCONRES    103
                       SCONRES     44
Joint Resolutions      HJRES      172
                       SJRES       69
Simple Resolutions     HRES       604
                       SRES       320

* 15 of 6726 bills are amendment document. (`EAH` and `EAS`)

In [206]:
df_bills.groupby(['Type', 'Subtype', 'Version Abbreviation', 'Version']).count()[['Filename']].rename(columns={'Filename': 'Count'})

Count
Type                   Subtype Version Abbreviation Version                            
Bills                  HR      EAH                  Engrossed Amendment House         2
                               EAS                  Engrossed Amendment Senate       10
                               EH                   Engrossed in House              300
                               ENR                  Enrolled Bill                    28
                               IH                   Introduced in House            3743
                               PCS                  Placed on Calendar Senate        18
                               RCS                  Reference Change Senate           1
                               RDS                  Received in Senate               28
                               RFS                  Referred in Senate              256
                               RH                   Reported in House               229
                               RS                   Reported in Senate               24
                       S       CPS                  Considered and Passed Senate      2
                               EAH                  Engrossed Amendment House         3
                               ENR                  Enrolled Bill                    11
                               ES                   Engrossed in Senate              71
                               IS                   Introduced in Senate           1787
                               PCS                  Placed on Calendar Senate        15
                               RCS                  Reference Change Senate           2
                               RFH                  Referred in House                40
                               RH                   Reported in House                 1
                               RS                   Reported in Senate              155
Concurrent Resolutions HCONRES EH                   Engrossed in House               10
                               ENR                  Enrolled Bill                     6
                               IH                   Introduced in House              73
                               RDS                  Received in Senate                8
                               RFS                  Referred in Senate                2
                               RH                   Reported in House                 4
                       SCONRES ATS                  Agreed to Senate                  5
                               ENR                  Enrolled Bill                     6
                               ES                   Engrossed in Senate               7
                               IS                   Introduced in Senate             18
                               PCS                  Placed on Calendar Senate         1
                               RFH                  Referred in House                 2
                               RS                   Reported in Senate                1
                               SC                   Sponsor Change                    4
Joint Resolutions      HJRES   EH                   Engrossed in House               19
                               ENR                  Enrolled Bill                    15
                               IH                   Introduced in House             116
                               PCS                  Placed on Calendar Senate         1
                               RDS                  Received in Senate               17
                               RFS                  Referred in Senate                2
                               RH                   Reported in House                 2
                       SJRES   CPS                  Considered and Passed Senate      1
                               ENR                  Enrolled Bill                     6
                               ES 

### Amendment

In [209]:
df_bills[df_bills['Version Abbreviation'].isin(['EAS', 'EAH'])].groupby(['Type', 'Subtype', 'Version Abbreviation', 'Version']).count()[['Filename']].rename(columns={'Filename': 'Count'})

Count
Type  Subtype Version Abbreviation Version                          
Bills HR      EAH                  Engrossed Amendment House       2
              EAS                  Engrossed Amendment Senate     10
      S       EAH                  Engrossed Amendment House       3

### 

In [210]:
df_bills[df_bills['Type'].isnull()]

Filename Number Subtype  \
519  92510.140382952224512.47152.xml    NaN     NaN   

                                                  Tags Type  \
519  [dc:publisher, paragraph, external-xref, form,...  NaN   

    Version Abbreviation Version Definition Chambers  
519                  NaN     NaN        NaN      NaN

In [225]:
df_bills.groupby(['Type', 'Subtype']).count()[['Filename']].rename(columns={'Filename': 'Count'})

Count
Type                   Subtype       
Bills                  HR        4639
                       S         2087
Concurrent Resolutions HCONRES    103
                       SCONRES     44
Joint Resolutions      HJRES      172
                       SJRES       69
Simple Resolutions     HRES       604
                       SRES       320

# 115 congress bills

In [321]:
# Bills
INDIR_115_HR = './data/115/bills/hr'
INDIR_115_S = './data/115/bills/s'

# Concurrent Resolutions
INDIR_115_HCONRES = './data/115/bills/hconres'
INDIR_115_SCONRES = './data/115/bills/sconres'

# Joint Resolutions
INDIR_115_HJRES = './data/115/bills/hjres'
INDIR_115_SJRES = './data/115/bills/sjres'

# Simple Resolutions
INDIR_115_HRES = './data/115/bills/hres'
INDIR_115_SRES = './data/115/bills/sres'

INDIR_ALL = [INDIR_115_HR, INDIR_115_S, 
              INDIR_115_HCONRES, INDIR_115_SCONRES, 
              INDIR_115_HJRES, INDIR_115_SJRES, 
              INDIR_115_HRES, INDIR_115_SRES
             ]

In [322]:
data = []

for i in INDIR_ALL:
    print('Processing {}'.format(i))
    for d in walk_dirs(i):
        data.append(d)
        
df_115 = pd.DataFrame(data)

Processing ./data/115/bills/hr
Processing ./data/115/bills/s
Processing ./data/115/bills/hconres
Processing ./data/115/bills/sconres
Processing ./data/115/bills/hjres
Processing ./data/115/bills/sjres
Processing ./data/115/bills/hres
Processing ./data/115/bills/sres


In [323]:
df_115

Directory   Number Subtype  \
0         ./data/115/bills/hr/hr1216/text-versions/ih   HR1216      HR   
1         ./data/115/bills/hr/hr337/text-versions/rfs    HR337      HR   
2          ./data/115/bills/hr/hr337/text-versions/ih    HR337      HR   
3          ./data/115/bills/hr/hr337/text-versions/eh    HR337      HR   
4         ./data/115/bills/hr/hr2493/text-versions/ih   HR2493      HR   
5         ./data/115/bills/hr/hr3541/text-versions/ih   HR3541      HR   
6         ./data/115/bills/hr/hr3773/text-versions/ih   HR3773      HR   
7          ./data/115/bills/hr/hr939/text-versions/ih    HR939      HR   
8         ./data/115/bills/hr/hr1024/text-versions/ih   HR1024      HR   
9          ./data/115/bills/hr/hr105/text-versions/ih    HR105      HR   
10        ./data/115/bills/hr/hr1818/text-versions/ih   HR1818      HR   
11        ./data/115/bills/hr/hr3317/text-versions/ih   HR3317      HR   
12        ./data/115/bills/hr/hr1440/text-versions/ih   HR1440      HR   
13         ./data/115/bills/hr/hr561/text-versions/ih    HR561      HR   
14        ./data/115/bills/hr/hr4132/text-versions/ih   HR4132      HR   
15        ./data/115/bills/hr/hr1672/text-versions/ih   HR1672      HR   
16         ./data/115/bills/hr/hr753/text-versions/ih    HR753      HR   
17        ./data/115/bills/hr/hr3919/text-versions/ih   HR3919      HR   
18        ./data/115/bills/hr/hr3125/text-versions/ih   HR3125      HR   
19        ./data/115/bills/hr/hr2231/text-versions/ih   HR2231      HR   
20         ./data/115/bills/hr/hr595/text-versions/ih    HR595      HR   
21        ./data/115/bills/hr/hr1686/text-versions/ih   HR1686      HR   
22        ./data/115/bills/hr/hr2003/text-versions/ih   HR2003      HR   
23        ./data/115/bills/hr/hr2467/text-versions/ih   HR2467      HR   
24        ./data/115/bills/hr/hr2655/text-versions/ih   HR2655      HR   
25        ./data/115/bills/hr/hr3787/text-versions/ih   HR3787      HR   
26        ./data/115/bills/hr/hr3926/text-versions/ih   HR3926      HR   
27       ./data/115/bills/hr/hr3328/text-versions/rfs   HR3328      HR   
28        ./data/115/bills/hr/hr3328/text-versions/ih   HR3328      HR   
29        ./data/115/bills/hr/hr3328/text-versions/rh   HR3328      HR   
...                                               ...      ...     ...   
9360   ./data/115/bills/sres/sres214/text-versions/is  SRES214    SRES   
9361  ./data/115/bills/sres/sres214/text-versions/ats  SRES214    SRES   
9362    ./data/115/bills/sres/sres78/text-versions/is   SRES78    SRES   
9363  ./data/115/bills/sres/sres283/text-versions/ats  SRES283    SRES   
9364  ./data/115/bills/sres/sres277/text-versions/ats  SRES277    SRES   
9365    ./data/115/bills/sres/sres47/text-versions/is   SRES47    SRES   
9366   ./data/115/bills/sres/sres47/text-versions/ats   SRES47    SRES   
9367  ./data/115/bills/sres/sres248/text-versions/ats  SRES248    SRES   
9368   ./data/115/bills/sres/sres270/text-versions/is  SRES270    SRES   
9369  ./data/115/bills/sres/sres284/text-versions/ats  SRES284    SRES   
9370   ./data/115/bills/sres/sres40/text-versions/ats   SRES40    SRES   
9371   ./data/115/bills/sres/sres246/text-versions/is  SRES246    SRES   
9372    ./data/115/bills/sres/sres49/text-versions/is   SRES49    SRES   
9373   ./data/115/bills/sres/sres82/text-versions/ats   SRES82    SRES   
9374   ./data/115/bills/sres/sres279/text-versions/is  SRES279    SRES   
9375   ./data/115/bills/sres/sres279/text-versions/rs  SRES279    SRES   
9376    ./data/115/bills/sres/sres76/text-versions/is   SRES76    SRES   
9377   ./data/115/bills/sres/sres76/text-versions/ats   SRES76    SRES   
9378  ./data/115/bills/sres/sres241/text-versions/ats  SRES241    SRES   
9379    ./data/115/bills/sres/sres71/text-versions/is   SRES71    SRES   
9380    ./data/115/bills/sres/sres85/text-versions/is   SRES85    SRES   
9381    ./data/115/bills/sres/sres25/text-versions/is   SRES25    SRES   
9382   ./data/115/bills/sres/sres215/text-versio

In [328]:
df_115[df_115['Version'] == 'N/A']

Directory  Number Subtype  Summary   Type Version
5372  ./data/115/bills/hr/hr4300  HR4300      HR        0  Bills     N/A
5373  ./data/115/bills/hr/hr4307  HR4307      HR        0  Bills     N/A
5374  ./data/115/bills/hr/hr4309  HR4309      HR        0  Bills     N/A
5375  ./data/115/bills/hr/hr4308  HR4308      HR        0  Bills     N/A
5376  ./data/115/bills/hr/hr4306  HR4306      HR        0  Bills     N/A
5377  ./data/115/bills/hr/hr4301  HR4301      HR        1  Bills     N/A
5378  ./data/115/bills/hr/hr4273  HR4273      HR        0  Bills     N/A
5379  ./data/115/bills/hr/hr4287  HR4287      HR        0  Bills     N/A
5380  ./data/115/bills/hr/hr4280  HR4280      HR        0  Bills     N/A
5381  ./data/115/bills/hr/hr4289  HR4289      HR        0  Bills     N/A
5382  ./data/115/bills/hr/hr4288  HR4288      HR        0  Bills     N/A
5383  ./data/115/bills/hr/hr4275  HR4275      HR        0  Bills     N/A
5384  ./data/115/bills/hr/hr4281  HR4281      HR        0  Bills     N/A
5385  ./data/115/bills/hr/hr4286  HR4286      HR        0  Bills     N/A
5386  ./data/115/bills/hr/hr4272  HR4272      HR        0  Bills     N/A
5387  ./data/115/bills/hr/hr4268  HR4268      HR        0  Bills     N/A
5388  ./data/115/bills/hr/hr4261  HR4261      HR        0  Bills     N/A
5389  ./data/115/bills/hr/hr4295  HR4295      HR        0  Bills     N/A
5390  ./data/115/bills/hr/hr4292  HR4292      HR        0  Bills     N/A
5391  ./data/115/bills/hr/hr4266  HR4266      HR        0  Bills     N/A
5392  ./data/115/bills/hr/hr4267  HR4267      HR        0  Bills     N/A
5393  ./data/115/bills/hr/hr4293  HR4293      HR        0  Bills     N/A
5394  ./data/115/bills/hr/hr4269  HR4269      HR        0  Bills     N/A
5395  ./data/115/bills/hr/hr4312  HR4312      HR        0  Bills     N/A
5396  ./data/115/bills/hr/hr4315  HR4315      HR        0  Bills     N/A
5397  ./data/115/bills/hr/hr4314  HR4314      HR        0  Bills     N/A
5398  ./data/115/bills/hr/hr4313  HR4313      HR        0  Bills     N/A
5399  ./data/115/bills/hr/hr4304  HR4304      HR        0  Bills     N/A
5400  ./data/115/bills/hr/hr4303  HR4303      HR        0  Bills     N/A
5401  ./data/115/bills/hr/hr4302  HR4302      HR        0  Bills     N/A
...                          ...     ...     ...      ...    ...     ...
5420  ./data/115/bills/hr/hr4316  HR4316      HR        0  Bills     N/A
5421  ./data/115/bills/hr/hr4311  HR4311      HR        0  Bills     N/A
5422  ./data/115/bills/hr/hr4310  HR4310      HR        0  Bills     N/A
5423  ./data/115/bills/hr/hr4317  HR4317      HR        0  Bills     N/A
7841    ./data/115/bills/s/s2100   S2100       S        0  Bills     N/A
7842    ./data/115/bills/s/s2106   S2106       S        0  Bills     N/A
7843    ./data/115/bills/s/s2101   S2101       S        0  Bills     N/A
7844    ./data/115/bills/s/s2089   S2089       S        0  Bills     N/A
7845    ./data/115/bills/s/s2087   S2087       S        0  Bills     N/A
7846    ./data/115/bills/s/s2080   S2080       S        0  Bills     N/A
7847    ./data/115/bills/s/s2086   S2086       S        0  Bills     N/A
7848    ./data/115/bills/s/s2088   S2088       S        0  Bills     N/A
7849    ./data/115/bills/s/s2095   S2095       S        0  Bills     N/A
7850    ./data/115/bills/s/s2092   S2092       S        0  Bills     N/A
7851    ./data/115/bills/s/s2093   S2093       S        0  Bills     N/A
7852    ./data/115/bills/s/s2094   S2094       S        0  Bills     N/A
7853    ./data/115/bills/s/s2104   S2104       S        0  Bills     N/A
7854    ./data/115/bills/s/s2103   S2103       S        0  Bills     N/A
7855    ./data/115/bills/s/s2102   S2102       S        0  Bills     N/A
7856    ./data/115/bills/s/s2105   S2105       S        0  Bills     N/A
7857    ./data/115/bills/s/s2083   S2083       S        0  Bills     N/A
7858    ./data/115/bills/s/s2070   S2070       S        0  Bills     N/A
7859    ./data/115/bills/s/s2084   S2084       S        0  Bills     N/A
7860    

* 9312 bills
* 6799/9312 with one summary

In [327]:
len(df_115[df_115['Version'] != 'N/A'])

9312

In [355]:
df_115[df_115['Version'] != 'N/A']['Summary'].value_counts()

1    6799
0    2513
Name: Summary, dtype: int64

In [367]:
# df_115[df_115['Version'] != 'N/A'].groupby(['Type'])['Summary'].agg(['count', 'sum']).rename(columns={'count': 'Count', 'sum': 'Summary'})
df_115[df_115['Version'] != 'N/A'].groupby(['Type'])['Summary'].agg(['count']).rename(columns={'count': 'Count (Version)'})

Count (Version)
Type                                   
Bills                              7789
Concurrent Resolutions              173
Joint Resolutions                   246
Simple Resolutions                 1104

In [368]:
# df_115[df_115['Version'] != 'N/A'].groupby(['Type', 'Subtype'])['Summary'].agg(['count', 'sum']).rename(columns={'count': 'Count', 'sum': 'Summary'})
df_115[df_115['Version'] != 'N/A'].groupby(['Type', 'Subtype'])['Summary'].agg(['count']).rename(columns={'count': 'Count (Version)'})

Count (Version)
Type                   Subtype                 
Bills                  HR                  5372
                       S                   2417
Concurrent Resolutions HCONRES              124
                       SCONRES               49
Joint Resolutions      HJRES                177
                       SJRES                 69
Simple Resolutions     HRES                 711
                       SRES                 393

In [346]:
len(df_number)

9312

In [347]:
df_number = df_number.drop_duplicates()

In [348]:
len(df_number)

7548

In [357]:
df_number['Summary'].value_counts()

1    5192
0    2356
Name: Summary, dtype: int64

In [369]:
df_number.groupby(['Type', 'Subtype'])['Summary'].agg(['count', 'sum']).rename(columns={'count': 'Count (Bill Number)', 'sum': 'Summary'})

Count (Bill Number)  Summary
Type                   Subtype                              
Bills                  HR                      4253     2844
                       S                       2070     1270
Concurrent Resolutions HCONRES                   91       85
                       SCONRES                   29       27
Joint Resolutions      HJRES                    120      116
                       SJRES                     49       48
Simple Resolutions     HRES                     611      534
                       SRES                     325      268